tests/unit/scanners/test.javascript.js

```diff
@@ -435,5 +457,36 @@ describe('JavaScript Scanner', () => {
 
       expect(jsScanner.sourceType).toEqual('script');
     });
+
+    it('returns the "parsing as a module" error', async () => {
+      const code = `
+        export const FOO = "FOO";
+        // This will produce a syntax error when parsing the file as module.
+        default FOO;
+      `;
+
+      const jsScanner = new JavaScriptScanner(code, 'code.js');
+      const detectedSourceType = jsScanner.detectSourceType('code.js');
+
+      expect(detectedSourceType.sourceType).toEqual('script');
+      expect(detectedSourceType.parsingError).toEqual({
+        type: 'module',
+        error: 'Unexpected token default at line: 4 and column: 9',
+      });
+    });
+
+    it('handles "parsing as a module" errors without line/column', async () => {
+      const code = `(function () {})();`;
+
+      const jsScanner = new JavaScriptScanner(code, 'code.js');
+      const detectedSourceType = jsScanner.detectSourceType('code.js');
+
+      expect(detectedSourceType.sourceType).toEqual('script');
+      expect(detectedSourceType.parsingError).toEqual({
+        type: 'module',
+        error: oneLine`Cannot read property 'type' of null at line: (unknown)
+          and column: (unknown)`,
+      });
+    });
```

Comment:
@willdurand while reading this test I was wondering "where is that error coming from?", and based on what I think is happening here it seems to me that this test case is actually leveraging a bug in our `JavascriptScanner`'s `_getSourceType` method:

https://github.com/mozilla/addons-linter/blob/5b722ec64f964d0bcc028dd7a32a3329e6c3444a/src/scanners/javascript.js#L207-L215

If I'm readin it correctly, that exception is likely coming from a call that goes into line 214, and it is due to a child node with a `null` value.

Looking to the AST that esprima should be returning to us on the code snippet `(function () {})();`:
- https://astexplorer.net/#/gist/6bdd2c8a13a959cc2527f1606ef073c2/64a9440ac8d1f15003339fba95fceb40653cfdcc

I think that the null node is actually the `id` property of the FunctionExpression node:
```
{
  type: 'FunctionExpression',
  id: null,
  ...
},
```

This looks like a bug in the `_getSourceType` logic that we should actually fix, e.g. by only recurse if the child node is defined and have the `type` property that we expect from any AST node.

As a side effect, that also means that we can't leverage it for this test, and so my second question is:
- is there any other condition (besides bug in our own code as in this case) that would generate a parsing error that doesn't have line and column properties? (I guess that another similar one may be bugs in eslint parsing itself, still not helpful for coverage).

I think that it is still reasonable to look into the behavior that we'd like the addons-linter to have in case of bugs like this one, but:
- to cover it with a test we may opt for temporarily overriding our `_getSourceType` method or `espree.parse` (e.g. using jest mocking features), this way the test should not break if the bug is being fixed
- we may want to evaluate if we can detect this conditions consistently and report the validation failure with a separate error code specific to "unexpected parsing issue that we would like to be reported to us" (because they are more likely to be bugs that we want to fix and not related to the limitation of the default eslint parser)

Let me know what do you think (and if you think that it would be useful to have a brief chat together over Matrix or Slack about this).

Reply:
> is there any other condition (besides bug in our own code as in this case) that would generate a parsing error that doesn't have line and column properties? (I guess that another similar one may be bugs in eslint parsing itself, still not helpful for coverage).

I am not so sure about that. That's the only error I found.